File: mobile/src/test_generator.py
"""
test_generator.py
Mobile equivalent of web/src/testGenerator.js

Responsibility:
  generate_test_cases(memory_log, output_path) — builds test case prompt,
  calls LLM in raw mode, parses JSON array response, saves to mobile_test_cases.json.
"""

import json
import os

from .mobile_preprocess import build_test_case_prompt
from .llm_client import call_llm


def generate_test_cases(memory_log: list, output_path: str) -> list:
    """
    Generates structured test cases from the completed exploration memory log.
    Mirrors generateTestCases() in testGenerator.js.

    Args:
        memory_log:  Full completed exploration log list
        output_path: Where to save mobile_test_cases.json

    Returns:
        Parsed and validated list of test case dicts (may be empty on LLM error)
    """
    if not memory_log:
        print("[test_generator] Memory log is empty — cannot generate test cases.")
        return []

    print(f"[test_generator] Building test case prompt from {len(memory_log)} steps...")
    prompt = build_test_case_prompt(memory_log)

    try:
        # IMPORTANT: response_format="raw" — the test-case response is a JSON
        # array, not a single exploration action, so it must NOT be routed
        # through parse_action() (which coerces everything into an
        # {action, elementId, resource_id, value, reason} dict).
        raw_response = call_llm(prompt, response_format="raw")
    except Exception as e:
        print(f"[test_generator] LLM call failed: {e}")
        return []

    if isinstance(raw_response, list):
        test_cases = raw_response
    else:
        test_cases = _parse_test_cases_json(str(raw_response))

    # Validate — each entry must have id, objective, steps, expected_result
    valid = []
    for tc in test_cases:
        if not isinstance(tc, dict):
            print(f"[test_generator] Skipping non-object test case entry: {str(tc)[:80]}")
            continue
        if tc.get("id") and tc.get("objective") and isinstance(tc.get("steps"), list) and tc.get("expected_result"):
            valid.append(tc)
        else:
            print(f"[test_generator] Skipping malformed test case: {str(tc)[:80]}")

    _save_test_cases(valid, output_path)

    print(f"\n[test_generator] Generated {len(valid)} test case(s):")
    for tc in valid:
        print(f"  [{tc['id']}] {tc['objective']} ({len(tc['steps'])} steps)")

    if not valid:
        print(
            "[test_generator] 0 test cases produced. If this wasn't a stub run, "
            "check that the LLM actually returned a JSON array matching the schema "
            "in build_test_case_prompt() — print raw_response above to inspect."
        )

    return valid


def _parse_test_cases_json(raw_str: str) -> list:
    """
    Attempts JSON.parse on raw LLM output.
    Strips markdown fences and retries once on failure.
    Mirrors _parseTestCasesJSON() in testGenerator.js.
    """
    # Attempt 1: direct parse
    try:
        parsed = json.loads(raw_str)
        return parsed if isinstance(parsed, list) else [parsed]
    except json.JSONDecodeError:
        pass

    # Attempt 2: strip markdown fences
    cleaned = raw_str.replace("```json", "").replace("```", "").strip()
    try:
        parsed = json.loads(cleaned)
        return parsed if isinstance(parsed, list) else [parsed]
    except json.JSONDecodeError as e:
        print(f"[test_generator] JSON parse failed after stripping fences: {e}")
        print(f"[test_generator] Raw response was: {raw_str[:500]}")
        return []


def _save_test_cases(test_cases: list, output_path: str) -> None:
    """Writes test case array to the given output path."""
    os.makedirs(os.path.dirname(output_path), exist_ok=True)
    with open(output_path, "w", encoding="utf-8") as f:
        json.dump(test_cases, f, indent=2)
    print(f"[test_generator] Saved -> {output_path}")

File: web/src/testGenerator.js
/**
 * testGenerator.js
 * Day 3 – Nidhi K
 * Responsibility:
 *   generateTestCases(memoryLog) — calls buildTestCasePrompt (from preprocess.js),
 *   sends to LLM via callLLM() (Navya's), parses the JSON array response,
 *   saves to logs/test_cases.json, and prints a summary.
 *
 * Depends on:
 *   - src/preprocess.js    (buildTestCasePrompt)
 *   - src/llmClient.js     (callLLM — implemented by Navya)
 */

const fs = require('fs');
const path = require('path');
const { buildTestCasePrompt } = require('./preprocess');
const { callLLM } = require('./llmClient');
const { reconstructWorkflows } = require('./exploreHelpers');

// Honors ARCH_A_OUTPUT_DIR so unified runs collect into runs/<id>/dom.
const OUTPUT_DIR = process.env.ARCH_A_OUTPUT_DIR || path.join(__dirname, '..', 'logs');
const TEST_CASES_PATH = path.join(OUTPUT_DIR, 'test_cases.json');

/**
 * generateTestCases(memoryLog)
 *
 * Builds prompt from the completed memory log, calls the LLM,
 * parses the returned JSON array of test cases, and saves to disk.
 *
 * Retry logic: if JSON.parse fails on first attempt, strips markdown
 * fences and retries once. If still fails, logs the raw response and
 * saves an empty array so the pipeline doesn't crash.
 *
 * @param {Array} memoryLog  - Full completed exploration log
 * @returns {Promise<Array>} - Parsed test case array (may be empty on LLM error)
 */
async function generateTestCases(input) {
  const memoryLog = Array.isArray(input) ? input : (input && input.memoryLog) || [];
  const transitions = (!Array.isArray(input) && input && input.transitions) || [];

  if (!memoryLog.length) {
    console.warn('[testGenerator] Memory log is empty — cannot generate test cases.');
    return [];
  }

  console.log(`[testGenerator] Building test case prompt from ${memoryLog.length} steps / ${transitions.length} transitions...`);
  const prompt = buildTestCasePrompt({ memoryLog, transitions });

  let rawResponse = null;
  try {
    rawResponse = await callLLM(prompt);
  } catch (err) {
    // LLM unavailable (rate limit, network, missing key) — fall through to
    // the deterministic grounded fallback instead of producing nothing.
    console.error('[testGenerator] LLM call failed:', err.message);
  }

  // rawResponse may be an already-parsed object (if callLLM returns object)
  // or a raw string. Normalise to string for parsing.
  let testCases = [];

  if (Array.isArray(rawResponse)) {
    // callLLM already parsed it
    testCases = rawResponse;
  } else if (rawResponse !== null && rawResponse !== undefined) {
    const rawStr = typeof rawResponse === 'string'
      ? rawResponse
      : JSON.stringify(rawResponse);

    testCases = _parseTestCasesJSON(rawStr);
  }

  // Validate structure — each entry must have id, objective, steps, expected_result
  testCases = testCases.filter(tc => {
    const valid = tc.id && tc.objective && Array.isArray(tc.steps) && tc.expected_result;
    if (!valid) {
      console.warn(`[testGenerator] Skipping malformed test case:`, JSON.stringify(tc).slice(0, 80));
    }
    return valid;
  });

  // Save to disk
  // Deterministic fallback: if the LLM produced nothing usable, reconstruct
  // test cases DIRECTLY from recorded successful workflows — 100% grounded,
  // zero extra LLM calls.
  if (!testCases.length && transitions.length) {
    console.warn('[testGenerator] No valid LLM test cases — using deterministic grounded fallback.');
    testCases = _deterministicTestCases(memoryLog, transitions);
  }

  _saveTestCases(testCases);

  // Print summary
  console.log(`\n[testGenerator] ✓ Generated ${testCases.length} test case(s):`);
  testCases.forEach(tc => {
    console.log(`  [${tc.id}] ${tc.objective} (${tc.steps.length} steps)`);
  });

  return testCases;
}

/**
 * _deterministicTestCases(memoryLog, transitions)
 * Reconstructs consecutive successful transitions into multi-step replayable
 * test cases. Every selector/URL/value comes verbatim from recorded history —
 * nothing can be invented.
 */
function _deterministicTestCases(memoryLog, transitions) {
  const workflows = reconstructWorkflows(transitions);

  return workflows.map((wf, i) => ({
    id: `TC${String(i + 1).padStart(3, '0')}`,
    objective: `Replay the recorded workflow (${wf.map(t => t.action.type).join(' → ')}), ending at ${wf[wf.length - 1].url_after}`,
    source: 'deterministic_fallback',
    steps: wf.map((t, j) => ({
      stepNum: j + 1,
      action: t.action.type,
      selector: t.action.target,
      value: t.action.value || '',
      description: `${t.action.type} ${t.action.target}${t.action.value ? ` with "${t.action.value}"` : ''}`,
    })),
    expected_result: `All ${wf.length} steps execute without errors; final page loads at ${wf[wf.length - 1].url_after}`,
  }));
}

/**
 * _parseTestCasesJSON(rawStr)
 * Attempts JSON.parse on raw LLM output.
 * Strips markdown code fences if present and retries once.
 *
 * @param {string} rawStr
 * @returns {Array}
 */
function _parseTestCasesJSON(rawStr) {
  // Attempt 1: direct parse
  try {
    const parsed = JSON.parse(rawStr);
    return Array.isArray(parsed) ? parsed : [parsed];
  } catch (_) {
    // Attempt 2: strip markdown fences (```json ... ```)
    const stripped = rawStr
      .replace(/```json\s*/gi, '')
      .replace(/```\s*/g, '')
      .trim();
    try {
      const parsed = JSON.parse(stripped);
      return Array.isArray(parsed) ? parsed : [parsed];
    } catch (err2) {
      console.error('[testGenerator] JSON parse failed after stripping fences:', err2.message);
      // Save raw response for debugging (inside this run's output dir)
      const debugPath = path.join(OUTPUT_DIR, 'test_cases_raw_response.txt');
      try {
        fs.mkdirSync(path.dirname(debugPath), { recursive: true });
        fs.writeFileSync(debugPath, rawStr, 'utf8');
        console.error(`[testGenerator] Raw LLM response saved to ${debugPath}`);
      } catch (_) {}
      return [];
    }
  }
}

/**
 * _saveTestCases(testCases)
 * Writes test case array to logs/test_cases.json.
 *
 * @param {Array} testCases
 */
function _saveTestCases(testCases) {
  const dir = path.dirname(TEST_CASES_PATH);
  fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(TEST_CASES_PATH, JSON.stringify(testCases, null, 2), 'utf8');
  console.log(`[testGenerator] Saved → ${TEST_CASES_PATH}`);
}

module.exports = { generateTestCases, _deterministicTestCases };
